x/ibc-transfer/keeper/relay.go

```diff
@@ -148,104 +116,96 @@ func (k Keeper) createOutgoingPacket(
-	coins := make(sdk.Coins, len(data.Amount))
-	for i, coin := range data.Amount {
-		if !strings.HasPrefix(coin.Denom, prefix) {
-			return sdkerrors.Wrapf(
-				sdkerrors.ErrInvalidCoins,
-				"%s doesn't contain the prefix '%s'", coin.Denom, prefix,
-			)
-		}
-		coins[i] = sdk.NewCoin(coin.Denom[len(prefix):], coin.Amount)
+	// mint new tokens if the source of the transfer is the same chain
+	if err := k.bankKeeper.MintCoins(
+		ctx, types.ModuleName, sdk.NewCoins(voucher),
+	); err != nil {
+		return err
 	}
 
-	// unescrow tokens
-	escrowAddress := types.GetEscrowAddress(packet.GetDestPort(), packet.GetDestChannel())
-	return k.bankKeeper.SendCoins(ctx, escrowAddress, receiver, coins)
+	// send to receiver
+	return k.bankKeeper.SendCoinsFromModuleToAccount(
+		ctx, types.ModuleName, receiver, sdk.NewCoins(voucher),
+	)
 }
 
 func (k Keeper) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, data types.FungibleTokenPacketData, ack types.FungibleTokenPacketAcknowledgement) error {
 	if !ack.Success {
-		return k.refundPacketAmount(ctx, packet, data)
+		return k.refundPacketToken(ctx, packet, data)
 	}
 	return nil
 }
 
 func (k Keeper) OnTimeoutPacket(ctx sdk.Context, packet channeltypes.Packet, data types.FungibleTokenPacketData) error {
-	return k.refundPacketAmount(ctx, packet, data)
+	return k.refundPacketToken(ctx, packet, data)
 }
 
-func (k Keeper) refundPacketAmount(ctx sdk.Context, packet channeltypes.Packet, data types.FungibleTokenPacketData) error {
+func (k Keeper) refundPacketToken(ctx sdk.Context, packet channeltypes.Packet, data types.FungibleTokenPacketData) error {
 	// NOTE: packet data type already checked in handler.go
 
-	if len(data.Amount) != 1 {
-		return sdkerrors.Wrapf(types.ErrOnlyOneDenomAllowed, "%d denoms included", len(data.Amount))
-	}
-
-	// check the denom prefix
-	prefix := types.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel())
-	source := strings.HasPrefix(data.Amount[0].Denom, prefix)
+	token := sdk.NewCoin(data.Denom, sdk.NewIntFromUint64(data.Amount))
 
 	// decode the sender address
 	sender, err := sdk.AccAddressFromBech32(data.Sender)
 	if err != nil {
 		return err
 	}
 
-	if source {
-		coins := make(sdk.Coins, len(data.Amount))
-		for i, coin := range data.Amount {
-			coin := coin
-			if !strings.HasPrefix(coin.Denom, prefix) {
-				return sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "%s doesn't contain the prefix '%s'", coin.Denom, prefix)
-			}
-			coins[i] = sdk.NewCoin(coin.Denom[len(prefix):], coin.Amount)
-		}
-
+	if types.SenderChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), token.Denom) {
 		// unescrow tokens back to sender
 		escrowAddress := types.GetEscrowAddress(packet.GetSourcePort(), packet.GetSourceChannel())
-		return k.bankKeeper.SendCoins(ctx, escrowAddress, sender, coins)
+		return k.bankKeeper.SendCoins(ctx, escrowAddress, sender, sdk.NewCoins(token))
```

Comment:
we don't need to update the token denom prefix here, right?

Reply:
correct. Refunds only occur on the sending side and the denom included in the fungible token packet is not prefixed so no changes need to be made. The denom should be the same as the one sent and escrowed or burnt

Reply:
awesome. I'm finally able to create a full mental model of the whole flow. Those prefixes were too confusing lol